File: Algorithm/bfs.py
from collections import deque

def bfs(adj, start):
    visited = [False] * len(adj)
    order = []
    queue = deque([start])
    visited[start] = True

    while queue:
        node = queue.popleft()
        order.append(node)

        for neighbor in adj[node]:
            if not visited[neighbor]:
                visited[neighbor] = True
                queue.append(neighbor)

    return order


def main():
    graph = [
        [1, 2],    # 0
        [0, 3, 4], # 1
        [0, 4],    # 2
        [1, 4],    # 3
        [1, 2, 3], # 4
    ]

    start_node = 0
    visit_order = bfs(graph, start_node)
    print(f"BFS visit order starting from node {start_node}: {visit_order}")


if __name__ == "__main__":
    main()

File: Algorithm/bfs.cpp
#include <iostream>
#include <queue>
#include <vector>

std::vector<int> bfs(const std::vector<std::vector<int>>& adj, int start) {
    std::vector<int> order;
    std::vector<bool> visited(adj.size());
    std::queue<int> q;

    visited[start] = true;
    q.push(start);

    while (!q.empty()) {
        int node = q.front();
        q.pop();
        order.push_back(node);

        for (int nbr : adj[node]) {
            if (!visited[nbr]) {
                visited[nbr] = true;
                q.push(nbr);
            }
        }
    }

    return order;
}

int main() {
    std::vector<std::vector<int>> graph = {
        {1, 2},    // 0
        {0, 3, 4}, // 1
        {0, 4},    // 2
        {1, 4},    // 3
        {1, 2, 3}  // 4
    };

    int start = 0;
    std::vector<int> visitOrder = bfs(graph, start);

    std::cout << "BFS visit order starting from node " << start << ": ";
    for (int node : visitOrder) {
        std::cout << node << " ";
    }
    std::cout << std::endl;

    return 0;
}
